### PR TITLE
Reapply "Mtp magic send (#1031)" (#1064)

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR Category

Distributed Strategy

### PR Types

Bug fixes 

### Description

Reapply "Mtp magic send (#1031)" (#1064)
This reverts commit 93776e3e16b4375b8a018807548e78cb97853fa7.

### 是否引起精度变化

否

### CODE REVIEW

**Review by From00 — APPROVED** [review]
LGTM

File: src/paddlefleet/models/gpt/mtp_embedding_layer.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                "MTPEmbeddingLayer should only be used when enable_mtp_magic_send=True"
+            )
+
+        global input_ids_for_mtp
+        if len(input_ids_for_mtp) == 0:
+            raise RuntimeError(
+                "input_ids_for_mtp deque is empty, broadcast may have failed"
+            )
+
+        input_ids = input_ids_for_mtp.popleft()
```
Comment (risemeup1111) [inline src/paddlefleet/models/gpt/mtp_embedding_layer.py:85]: P1 **优先级：P1**

这里新增了 magic send 的唯一消费端，但当前 PR 没有任何生产路径向 `input_ids_for_mtp` 写入数据。全仓搜索只有测试里直接 `append`，`broadcast_data_obj` / `init_magic_send_comm_group` 也没有被调用；因此只要通过 `gpt_builder` 打开 `enable_mtp_magic_send=True`，`MTPEmbeddingLayer.forward()` 第一次执行就会走到这里抛出 `input_ids_for_mtp deque is empty`，功能无法跑通。

请把 dataloader/PP micro-batch 生产侧接进来，在首 stage 将每个 micro-batch 的 `input_ids` 广播到 magic-send group，并在消费 stage 按 forward 顺序入队；或者避免全局队列，直接把广播后的 `input_ids` 放进 `dict_args` 传给本层。实现形态至少需要类似：

```python
if config.enable_mtp_magic_send:
    group = init_magic_send_comm_group()
    mtp_input_ids = broadcast_data_obj(
        batch["input_ids"],
        src_rank=first_pp_rank,
        group=group,
    )
    if is_magic_send_consumer_rank:
        input_ids_for_mtp.append(mtp_input_ids)
```

File: src/paddlefleet/models/gpt/gpt_model.py [hunk 2]
```diff
@@ -279,6 +292,18 @@ def get_layer_desc_list(self, spec, tie_word_embeddings):
             )
 
         if spec.mtp:
+            # MTP magic send: MTPEmbeddingLayer shares weight with GPTEmbedding
+            # via SharedLayerDesc("mtp_embed") broadcast mechanism.
+            if spec.mtp_embedding:
+                self.add_sequential_layer(
+                    layers,
+                    SharedLayerDesc(
+                        "mtp_embed",
```
Comment (risemeup1111) [inline src/paddlefleet/models/gpt/gpt_model.py:301]: P1 **优先级：P1**

当 `tie_word_embeddings=True` 且 PP>1 时，首层 embedding 在上面的分支被注册到共享组 `"embed"`，但这里 MTPEmbeddingLayer 总是注册到 `"mtp_embed"`。两个 `SharedLayerDesc` 名称不同就不会共享参数；而 `MTPEmbeddingLayer` 里还把初始化关掉了，所以 magic-send 的重嵌入会使用未同步/零初始化的权重，MTP logits 会错。

请在 tying 场景复用 `"embed"`，非 tying 场景才使用 `"mtp_embed"`：

```suggestion
                        "embed" if tie_word_embeddings else "mtp_embed",
```

File: src/paddlefleet/transformer/multi_token_prediction.py [hunk 3]
```diff
@@ -683,6 +686,159 @@ def forward(self, dict_args: dict):
...
+            # Concat [backbone_hidden | mtp_hidden] for unified split in downstream LM heads.
+            backbone_hs = dict_args.get(
+                "_backbone_hidden_states", hidden_states
+            )
+            dict_args["hidden_states"] = paddle.concat(
+                [backbone_hs, hidden_states]
+            )
+
+            # Strip auxiliary float tensors (stop_gradient=False) to avoid grad=None crash in P2P backward.
+            _keep_keys = {"hidden_states", "labels"}
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/multi_token_prediction.py:834]: P2 **优先级：P2**

这里清理 `dict_args` 时把 `blocks` 也删掉了。`TransformerLayer` 在 `block_attention_residuals=True` 时会把 completed blocks 保存在 `dict_args["blocks"]`，后面的 `GPTLMHead` / `GPTMainLMHead` 需要它们计算最终 BlockAttnRes；magic-send 路径只保留 `hidden_states` / `labels` 后，LM head 会退化成空 blocks，输出与非 magic-send 路径不一致。

请保留 `blocks`（或在不支持该组合时在配置校验中显式拒绝）：

```suggestion
            _keep_keys = {"hidden_states", "labels", "blocks"}
```

File: src/paddlefleet/pipeline_parallel/pp_utils/pp_comm_utils.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        list: one chunk
+    """
+    ret = []
+    while grouped:
+        if sum([r[1].size() for r in ret]) > split_size:
+            yield ret
+            ret = []
+        ret.append(grouped.pop())
+    if ret:
+        yield ret
```
Comment (risemeup1111) [inline src/paddlefleet/pipeline_parallel/pp_utils/pp_comm_utils.py:63]: P2 **优先级：P2**

`split_group` 在加入当前元素之前才检查 `sum(ret) > split_size`，所以第二个元素会先被塞进当前 chunk，导致 chunk 总大小已经超过阈值才在下一轮 yield；这和“避免单次 concat 超限”的目标不一致，`broadcast_data_obj` 仍可能构造过大的 buffer。

请在追加元素前用 `ret_size + item_size` 判断是否需要切 chunk：

```suggestion
    ret = []
    ret_size = 0
    while grouped:
        item = grouped.pop()
        item_size = item[1].size()
        if ret and ret_size + item_size > split_size:
            yield ret
            ret = []
            ret_size = 0
        ret.append(item)
        ret_size += item_size
    if ret:
        yield ret
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
发现几个需要先修正的问题，具体细节已经放在行级评论里。

- P3 **优先级：P3** 非行级：PR 描述/CI 模板项无法附着到 diff 行。`Check PR Description` 当前失败，且描述里的 `PR Category` 仍为空、测试/风险说明也未补充；建议按模板补齐类别，并说明 magic-send 场景的验证方式和兼容限制。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by Hz188 — APPROVED** [review]
